linux_os/guide/system/software/integrity/crypto/harden_sshd_macs_openssh_conf_crypto_policy/oval/shared.xml

```diff
@@ -0,0 +1,35 @@
+{{%- set PATH = "/etc/crypto-policies/back-ends/openssh.config" -%}}
+<def-group>
+  <definition class="compliance" id="{{{ rule_id }}}" version="1">
+    {{{ oval_metadata("Limit the Message Authentication Codes (MACs) to those which are FIPS-approved.") }}}
+    <criteria operator="AND" comment="Test conditions - presence of the file plus.">
+      <criterion comment="Check that {{{ PATH }}} contains FIPS-approved SSHD MACs" test_ref="test_{{{ rule_id }}}" />
+    </criteria>
+  </definition>
+
+  <ind:textfilecontent54_test check="all"
+  comment="test the value of MACs setting in the {{{ PATH }}} file"
+  id="test_{{{ rule_id }}}" version="1">
+    <ind:object object_ref="obj_{{{ rule_id }}}" />
+    <ind:state state_ref="ste_{{{ rule_id }}}" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object id="obj_{{{ rule_id }}}" version="1">
+    <ind:filepath>{{{ PATH }}}</ind:filepath>
+    <ind:pattern operation="pattern match">^MACs.*$</ind:pattern>
+    <ind:instance operation="greater than or equal" datatype="int">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+  <ind:textfilecontent54_state id="ste_{{{ rule_id }}}" version="1">
+    <ind:text var_ref="sshd_macs_crypto" operation="equals"></ind:text>
+  </ind:textfilecontent54_state>
+
+  <local_variable id="sshd_macs_crypto" datatype="string" comment="The regex of the directive" version="1">
+    <concat>
+      <literal_component>MACs </literal_component>
+      <variable_component var_ref="sshd_approved_macs"/>
```

Comment:
This OVAL check doesn't allow a subset of the algorithms. The implementation for the sshd_config file accounts for that and that's why we have decided to use a hardcoded value for the regex:

https://github.com/ComplianceAsCode/content/blob/602a7e283c9a03e56c1921b76312cd86c16478be/linux_os/guide/services/ssh/ssh_server/sshd_use_approved_macs_ordered_stig/oval/shared.xml#L34

From [V-230251](https://vaulted.io/library/disa-stigs-srgs/red_hat_enterprise_linux_8_security_technical_implementation_guide/V-230251)
`
If the MAC entries in the "openssh.config" and "opensshserver.config" files have any hashes other than "hmac-sha2-512" and "hmac-sha2-256", the order differs from the example above, if they are missing, or commented out, this is a finding.
`
For example if someone configures something `MACs hmac-sha2-512` (which is valid from a STIG standpoint), the check will fail.

The only way to solve this is to have a specialized STIG rule, unfortunately. Unless we are ok with our check forcing to always require both values.

Reply:
Could we hoist regex creation into a Python-based jinja function (see `add_python_functions` in `ssg/jinja.py`)? There's been enough times that we've wanted to do _something_ along these lines (perhaps ordered or unordered subset) that perhaps its time to expand the helpers and turn this into a safe-ish (either handling escaping or explicitly warning about bad characters) Python helper. 

Thoughts?

Reply:
Concretely, in the case where we're comparing against a variable, we'd need to know the order of things in the variable and possible entries. Then you can use a regex to select each portion (splitting on a comma) across both file contents and variable contents and then drop them into a `set` and use `subset of` as the comparison operator? Just a thought.

Reply:
We have discussed these possibilities in the past. The problem with variable in this case is that it requires ordering and OVAL cannot guarantee that. I think we cannot achieve this, but maybe I'm wrong.

Reply:
> This OVAL check doesn't allow a subset of the algorithms. The implementation for the sshd_config file accounts for that and that's why we have decided to use a hardcoded value for the regex:
> 
> https://github.com/ComplianceAsCode/content/blob/602a7e283c9a03e56c1921b76312cd86c16478be/linux_os/guide/services/ssh/ssh_server/sshd_use_approved_macs_ordered_stig/oval/shared.xml#L34
> 
> From [V-230251](https://vaulted.io/library/disa-stigs-srgs/red_hat_enterprise_linux_8_security_technical_implementation_guide/V-230251)
> `If the MAC entries in the "openssh.config" and "opensshserver.config" files have any hashes other than "hmac-sha2-512" and "hmac-sha2-256", the order differs from the example above, if they are missing, or commented out, this is a finding.`
> For example if someone configures something `MACs hmac-sha2-512` (which is valid from a STIG standpoint), the check will fail.
> 
> The only way to solve this is to have a specialized STIG rule, unfortunately. Unless we are ok with our check forcing to always require both values.

So the way I interpret this is: `Make sure the only value in MACs is <exactly what we tell you it should be (ie, hmac-sha2-512,hmac-sha2-256)`. This is why using the variable makes sense because we don't need to check for anything else other than what DISA wants to see. There is no subset check because that falls under DISA's `if they are missing - this is a finding` statement.

Reply:
I hadn't known about the "the order differs from the example above" -- so I am OK with a strict equality check.